bpf/process/types/basic.h

```diff
@@ -697,26 +695,41 @@ struct file_buf {
  * a reverse search.
  */
 static inline __attribute__((always_inline)) long
-filter_file_buf(struct file_buf *filter, char *args)
+filter_file_buf(struct file_buf *filter, struct string_buf *args)
 {
 	char *values = (char *)filter + sizeof(struct file_buf); // all values are after the header
-	__u32 remain = filter->total_sz - sizeof(__u32) - sizeof(__u32);
-	int i, next;
+	__u32 next, remain = filter->total_sz - sizeof(__u32) - sizeof(__u32);
+	long i, match;
+
+	/* There are cases where file pointer may not contain a path.
+	 * An example is using an unnamed pipe. This is not a match.
+	 */
+	if (args->len == 0)
+		return 0;
 
 #ifndef __LARGE_BPF_PROG
 #pragma unroll
 #endif
 	for (i = 0; i < MAX_MATCH_FILE_VALUES; ++i) {
-		next = __filter_file_buf((struct string_buf *)values, (struct string_buf *)args, filter->op);
-		if (!next)
+		struct string_buf *value = (struct string_buf *)values;
+
+		match = __filter_file_buf(value, args, filter->op);
+		if (filter->op == op_filter_neq || filter->op == op_filter_str_notprefix || filter->op == op_filter_str_notpostfix) {
+			if (match != FILTER_FILE_NOMATCH)
+				return 0;
+		} else if (match == FILTER_FILE_MATCH) {
 			return 1;
+		}
```

Comment:
I'd personally write this as:

```
....
// the file matched, which means that we can now make a decision.
if (match == FILTER_FILE_MATCH) {
     return is_not_operator(filter->op) ? 0 : 1; 
}
```

Where:
```
bool is_not_operator(op) {
    return (filter->op == op_filter_neq || filter->op == op_filter_str_notprefix || filter->op == op_filter_str_notpostfix);
}
```

But that's just me :)

Reply:
Simplifies the code a bit. Done!